### PR TITLE
BUG: Cope with indirect objects in filters and remove deprecated code

### LINKED ISSUE #2158: FlateDecode will fail if `columns` are an IndirectObject
Retrieving the images from a PDF file might fail for the FlateDecode case when `decode_parms` is something like

> {'/Colors': 3, '/Columns': IndirectObject(33, 0, 140029965619504), '/Predictor': 15}

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.14.21-150400.24.81-default-x86_64-with-glibc2.31

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==3.15.5, crypt_provider=('pycryptodome', '3.18.0'), PIL=10.0.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader


PATH = 'file.pdf'
for index, page in enumerate(PdfReader(PATH).pages):
    print(index, page)
    for key in page.images.keys():
        print(key)
        print(page.images[key].indirect_reference)
        page.images[key].image.convert('RGB').save(key[1:] + '.png')
```

## Traceback

This is the complete traceback I see:

```
Traceback (most recent call last):
  File "/home/stefan/temp/run.py", line 9, in <module>
    print(page.images[key].indirect_reference)
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/_page.py", line 2637, in __getitem__
    return self.get_function(index)
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/_page.py", line 544, in _get_image
    imgd = _xobj_to_image(cast(DictionaryObject, xobjs[id]))
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/filters.py", line 988, in _xobj_to_image
    data = x_object_obj.get_data()  # type: ignore
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/generic/_data_structures.py", line 953, in get_data
    decoded.set_data(b_(decode_stream_data(self)))
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/filters.py", line 696, in decode_stream_data
    data = FlateDecode.decode(data, params)
  File "/home/stefan/temp/venv/lib/python3.9/site-packages/pypdf/filters.py", line 166, in decode
    math.ceil(columns * colors * bits_per_component / 8) + 1
TypeError: unsupported operand type(s) for *: 'IndirectObject' and 'NumberObject'
```

### LINKED ISSUE #2159: Possibly undefined variable in FlateDecode.decode if parameters are an ArrayObject
When looking through the code for understanding #2158, I stumbled upon some possibly undefined variable.

Let's take the following snippet: https://github.com/py-pdf/pypdf/blob/05f2a651917e1bd2fed6634efda6bb345b31de6b/pypdf/filters.py#L143-L165

It seems like in the first branch (lines 144 to 150), `colors` will never be defined, thus possibly leading to a crash at line 164 if I did not miss something. (I do not have any PDF file for this, as I just stumbled upon this by accident.)

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2158
closes #2159

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma the coverage analysis failed. Can you rerun-it?

**Comment by MartinThoma** [comment]
The PR overall increased coverage, but this is not covered:

image

**Comment by MartinThoma** [comment]
I'll take care of that. I don't think we should cover that

File: pypdf/filters.py [hunk 1]
```diff
@@ -600,13 +587,16 @@ def _get_parameters(
...
+        decode_parms: Optional[DictionaryObject] = None,
         height: int = 0,
         **kwargs: Any,
     ) -> bytes:
+        # decode_parms is unused here
         if "decodeParms" in kwargs:  # deprecated
             deprecate_with_replacement("decodeParms", "parameters", "4.0.0")
             decode_parms = kwargs["decodeParms"]
+        if isinstance(decode_parms, ArrayObject):  # type: ignore
+            raise DeprecationError("Decode_params as ArrayObject is depreciated")
```
Comment (MartinThoma) [inline pypdf/filters.py:599]:
```suggestion
        if isinstance(decode_parms, ArrayObject):  # deprecated
            raise DeprecationError("Decode_params as ArrayObject is depreciated")
```

File: pypdf/filters.py [hunk 2]
```diff
@@ -600,13 +587,16 @@ def _get_parameters(
...
+        decode_parms: Optional[DictionaryObject] = None,
         height: int = 0,
         **kwargs: Any,
     ) -> bytes:
+        # decode_parms is unused here
         if "decodeParms" in kwargs:  # deprecated
             deprecate_with_replacement("decodeParms", "parameters", "4.0.0")
             decode_parms = kwargs["decodeParms"]
+        if isinstance(decode_parms, ArrayObject):  # deprecated
+            raise DeprecationError("Decode_params as ArrayObject is depreciated")
```
Comment (MartinThoma) [inline pypdf/filters.py:599]:
```suggestion
            deprecation_no_replacement("decode_parms being an ArrayObject", removed_in="3.15.5")
```

**Comment by MartinThoma** [comment]
```
from ._utils import deprecation_no_replacement
```

is missing

**Pull request merged**